odxtools/statemachine.py

```diff
@@ -145,6 +145,18 @@ def execute(self, service: "DiagService", **service_params: Any
         # ask the calling code to send the request to the ECU and
         # report back the reply
         raw_resp = yield raw_req
+
+        decoded_req_params = service.request.decode(raw_req)
```

Comment:
Why decode request if you already have `service_params` ?

Reply:
because optional parameters don't need to be specified for `service_params`. (this is a kludge to ensure that the same parameters as seen by the target ECU are used.)

Reply:
but optional (constant) parameters are not allowed in transitions as well.

Reply:
as far as I understood this, they are allowed. What's not allowed is specifying a value for them, but they can be tested for their presence. Also, value parameters can be defaulted and thus do not need to be specified for `encode()`. (Defaulted parameters that have not been specified for `encode()` could be handled here, but keeping this exactly equivalent to the encoding code is probably quite a headache which can be avoided by simply decoding the raw message, in particular since this code is not performance critical...)